### ISSUE TITLE
finger tutorial uses unclear python idioms.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2110]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)| @jerub reported|
|-|-|
|Trac ID|trac#2110|
|Type|defect|
|Created|2006-09-20 06:11:00Z|
|Branch|https://github.com/twisted/twisted/tree/finger-python-idioms-2110|

Here is one idiom that should be broken down:
```
        f.protocol,f.nickname,f.getUser = IRCReplyBot,nickname,self.getUser
```
into:
```
        f.protocol = IRCReplyBot
        f.nicname = nickname
        f.getUser = self.getUser
```

This violates our style guide (no simple suites):
```
    def connectionLost(self,reason): self.factory.setUser(*self.lines[:2])
```

This could be written better:
```
        self.factory.getUser(user
        ).addErrback(lambda _: "Internal error in server"
        ).addCallback(lambda m:
                      (self.transport.write(m+"\r\n"),
                       self.transport.loseConnection()))
```
Perhaps as:
```
        d = self.factory.getUser(user)

        def onError(err):
            return 'Internal error in server'
        d.addErrback(onError)

        def writeResponse(message)
            self.transport.write(message + '\r\n')
            self.transport.loseConnection()
        d.addCallback(writeResponse)
```
Or something similarly explicit.

Attachments:

* [finger-listings.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/086/086e894abe76b5dfda9ebd50da351e1d31d29c97/ff8194c68b7689c401a64d0f6ed6d049690563b7.diff) (19659 bytes) - added by kehander1 on 2008-10-23 16:58:48Z - Cleans up finger11.tac - finger18.tac
* [finger-listings-v2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/086/086e894abe76b5dfda9ebd50da351e1d31d29c97/e84c490741abb630b8db9b685194f6176445a717.diff) (19214 bytes) - added by kehander1 on 2008-10-24 17:11:45Z - Cleans up finger11.tac - finger18.tac
* [finger-listings-from-2110.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/086/086e894abe76b5dfda9ebd50da351e1d31d29c97/b7ac1efcb4c63664d5f4ddd6036de66e8cf5d341.diff) (4858 bytes) - added by kehander1 on 2008-10-27 19:14:43Z - Changes to finger12.tac - finger17.tac from branch
* [finger-listings-v3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/086/086e894abe76b5dfda9ebd50da351e1d31d29c97/a3b10c896b5bc3c6d5aa99076fd071b8c1d9e2d2.diff) (7578 bytes) - added by kehander1 on 2008-12-01 22:49:38Z - More changes to branch for finger12.tac - finger17.tac

<details><summary>Searchable metadata</summary>

```
trac-id__2110 2110
type__defect defect
reporter__jerub jerub
priority__lowest lowest
milestone__ 
branch__branches_finger_python_idioms_2110 branches/finger-python-idioms-2110
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1158732660000000 1158732660000000
changetime__1231264346000000 1231264346000000
version__None None
owner__ 
cc__thijs cc__kehander1
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_1"></a>|@jerub commented|
|-|-|

Specific url in querstion is: http://twistedmatrix.com/projects/core/documentation/howto/tutorial/protocol.html

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jerub<br>@jerub set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [24531]) Branching to 'finger-python-idioms-2110'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

Here's a suggestion from [#2281](https://github.com/twisted/twisted/issues/2281), let me know if this idiom should be replaced like that as well:

----
17 - Announce on the Web, Too
This is one of those showoff programs that adds a web server just because it
can.  

To do this, the service defines a method 'getResource'.  It returns a generic
'twisted.web.resource.Resource' and sets a function attribute 'getChild'.
Again the author uses Python shortcuts, an orthodox version would look like:
```python
    from twisted.web import resource, server, static
    class MyResource(resource.Resource):
        def getChild(self, path, request):
            """'path' is a string. 
                'request' is a 'twisted.protocols.http.Request'.
            """
            user = path
            value = self.users.get(user, 
                "No such user.<p/>http://this.site/user")
            contentType = "text/html"
            value = cgi.excape(value)
            contentType = cgi.escape(contentType)
            output = "<h1>%s</h1><p>%s</p>" % (value, contentType)
            return static.Data(output)
    class FingerService(service.Service):
        ....
        def getResource(self):
            return MyResource()
```
I don't know why the original wants to display "text/html" literally,
but that's what it does.

The application then opens a TCPServer listening on port 8000 for the web 
service.

'FingerService.__init__' and '_read' are slightly different than in the previous
example.  The previous version never clears out the 'users' dict after
instantiation; it only adds or replaces entries.  This is probably bad because
obsolete entries will remain.  The current example rectifies this by clearing
out 'users' before each refresh.  If 'users' were being shared in a 
multithreaded program we'd need a lock to prevent another thread from reading
it while the database is being populated.  But since this program is not
multithreaded we don't have to worry about that. 
----

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

Another suggestion from [#2281](https://github.com/twisted/twisted/issues/2281):

----
14 - Setting Message By Local Users
This program has two protocol-factory-TCPServer pairs, which are both child
services of the application.

The FingerSetterFactory is initialized with the FingerFactory instance so it
can access the 'users' dictionary.  The factory uses Python shortcuts.  An
equivalent orthodox factory would look like this:
```python
    from twisted.internet import protocol
    class FingerSetterFactory(protocol.ServerFactory):
        def __init__(self, ff):
            self.ff = ff
        def setUser(self, user, value):
            self.ff.users[user] = value

'FingerProtocol.connectionList' also uses Python shortcuts.  An equivalent
orthodox method is:
    def connectionLost(self, reason):
        user = self.lines[0]
        value = self.lines[1]
        self.factory.setUser(user, value)
```
----

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

Another one from [#2281](https://github.com/twisted/twisted/issues/2281):
----
8 - Output from Non-empty Factory
'FingerFactory' uses some coding shortcuts.  The obvious one is that 
usernames/output text are defined via keyword arguments.  The less obvious one
is that 'FingerFactory.getUser' generates the output text via Python's
 standard 'dict.get' method, which happens to let us define the "Not found"
value in the same step.
----

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set owner to @radix|
|-|-|

1.

```
-    def __init__(self, ff): self.setUser = ff.users.__setitem__
 
+    def __init__(self, ff):
+        self.setUser = ff.users.__setitem__
+
```

This is an improvement, but it could be better :) 

 - setUser should be explicitly defined as a method which does "self.users[name] = value"
 - 'ff' should be renamed to 'fingerFactory'

2. It looks like there are a bunch of tab characters in these files, some of them new. Can you replace all of them with spaces? (presumably with a sed script or something)

I guess I should verify the correctness of all the .tac files manually, since there are no unit tests for them. Man, what a pain. I don't have the time to do that right now, so I can't give this branch a +1 yet. Bug me on IRC once you've fixed these issues and I'll try to find some time to run through them.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @thijstriemstra|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|kehander1 removed owner|
|-|-|

This almost makes finger18.tac (and by extension, tutorial/style.xhtml) obsolete, since finger17.tac and everything before it now uses the same code as finger18.tac.  The only thing I can't figure out is what to do with that getChild lambda function in finger16.tac and finger17.tac .

As per [#2281](https://github.com/twisted/twisted/issues/2281), having two versions of the same code is confusing.

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|kehander1 commented|
|-|-|

Those superfluous inclusions of FingerSetterProtocol should go too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

It looks like finger-listings-v2.diff is against trunk, but it includes most of the changes from the branch plus a couple new ones.  If the idea is to add new changes to the branch, the best idea is to submit a diff against the branch instead.  Either way, always be sure to say what a diff is against. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Hm.  I wonder why finger18.tac dropped the use of `FingerSetterProtocol`.  Probably an accident.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [25091]) Apply the second half of that one patch that guy said

refs [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

Removing the review keyword cause it wasn't exarkun that added this which made the last changes to this ticket, so not sure if it should be up for review yet.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|kehander1 set owner to kehander1|
|-|-|

All right, this patch diffs from branches/finger-python-idioms-2110, and just incorporates some of the above suggestions, renames "Motd" to something slightly more consistent, and removes those FingerSetterProtocol bits.

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kehander1 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [25500]) Apply finger-listings-from-2110.diff

refs [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set owner to kehander1|
|-|-|

  1. The first point from [#comment:10 radix's review] doesn't seem to have been addressed.  Oh.  The second point hasn't been addressed either (finger13.tac, at least, still has a tab).  Also, there's some trailing whitespace in finger13.tac.
  1. The change to finger14.tac to remove the `users` attribute from `FingerService` seems incorrect.
  1. The docstring added to finger15.tac isn't indented properly and there's a bunch of trailing whitespace in this file.
  1. finger16.tac and finger17.tac has trailing whitespace and tabs

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kehander1 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|kehander1 commented|
|-|-|

(Diffed from 25500)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [25832]) Apply finger-python-idioms-2110

refs [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [25833]) Fix invalid syntax

refs [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Some more minor fixes in r25834 and r25835.  This seems like it's all improvements now (although the finger tutorial could still use a lot _more_ improvements).  So I'm going to merge the branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set status to `closed`|
|-|-|

(In [25837]) Merge finger-python-idioms-2110 

Author: kehander1
Reviewer: radix, thijs, exarkun, itamarst
Fixes: [#2110](https://github.com/twisted/twisted/issues/2110)

Remove ridiculous, obscure Python tricks from most of the finger tutorial listings;
replace them with straightforward, easy to understand idioms.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
